### PR TITLE
feat: implement repo-mode

### LINKED ISSUE #218: Implement "repo-mode" to use the repo as source of truth for devcontainer/Dockerfile
As part of #128, we want to implement "repo"-mode for envbuilder as an alternative to the current "filesystem"-mode.

Filesystem-mode work like this:
- Check if repository is cloned
- If no, clone the repository to the filesystem
- If yes, skip clone
- Read devcontainer/Dockerfile from filesystem

In that last step, the devcontainer or Dockerfile may be old or locally modified. It depends on what the user has done.

Repo-mode differs from filesystem-mode in that it will always clone the repo (to a temporary location) and use the devcontainer/Dockerfile from there. This way the resulting container is always in-sync with the repo.

For now, we can keep this simple and always clone the repo. This feature should be implemented as a package that can be used by both envbuilder and a future envbuilder Terraform provider.

In the future, we can improve performance by only reading the relevant files from the repo.

### PR DESCRIPTION
**Pull request by mafredri**
Fixes #218

### CODE REVIEW

File: git/git.go [hunk 1]
```diff
@@ -253,3 +291,80 @@ func SetupRepoAuth(options *options.Options) transport.AuthMethod {
 	}
 	return auth
 }
+
+func CloneOptionsFromOptions(options options.Options) (CloneRepoOptions, error) {
```
Comment (johnstcn) [inline git/git.go:295]: 👍

File: git/git.go [hunk 2]
```diff
@@ -253,3 +291,80 @@ func SetupRepoAuth(options *options.Options) transport.AuthMethod {
...
+func (w *progressWriter) Close() error {
+	err := w.r.Close()
+	err2 := w.WriteCloser.Close()
+	if err != nil {
+		return err
+	}
+	return err2
+}
+
+func ProgressWriter(write func(line string)) io.WriteCloser {
```
Comment (johnstcn) [inline git/git.go:347]: nice!

File: options/testdata/options.golden [hunk 3]
```diff
@@ -138,6 +138,14 @@ OPTIONS:
           Push the built image to a remote registry. This option forces a
           reproducible build.
 
+      --repo-build-mode local|remote, $ENVBUILDER_REPO_BUILD_MODE (default: local)
+          The mode to use when building the workspace from a repository. When
+          set to `local`, the repository is cloned locally and built. If the
+          repository is already present, it is used as the source of truth. When
```
Comment (johnstcn) [inline options/testdata/options.golden:144]: nit: add some explicit statement about what happens with local changes here

File: envbuilder.go [hunk 4]
```diff
@@ -182,6 +135,40 @@ func Run(ctx context.Context, opts options.Options) error {
...
 		}
+
+		// The repo wasn't cloned (i.e. may not be up-to-date with remote), so
+		// we need to clone it to get the right build context. This is only
+		// necessary when the repo is in remote build mode. If the repo is in
+		// local build mode, the build context is already available on the host
+		// filesystem.
+		//
+		// Skipping clone here if we believe workspace repo matches remote repo
+		// is a performance optimization.
```
Comment (johnstcn) [inline envbuilder.go:146]: is this a TODO?

Comment (mafredri) [reply envbuilder.go:146]: This was my attempt at explaining why we have the if in relation to the previous clone of the users workspace repo. Not a todo but wording can perhaps be improved. Or we can ditch it and always clone.

File: options/options.go [hunk 5]
```diff
@@ -417,6 +431,18 @@ func (o *Options) CLI() serpent.OptionSet {
...
+			Flag:    "repo-build-mode",
+			Env:     WithEnvPrefix("REPO_BUILD_MODE"),
+			Value:   serpent.EnumOf(&o.RepoBuildMode, RepoBuildModeLocal, RepoBuildModeRemote),
+			Default: RepoBuildModeLocal,
+			Description: "The mode to use when building the workspace from a " +
+				"repository. When set to `local`, the repository is cloned locally " +
+				"and built. If the repository is already present, it is used as the " +
+				"source of truth. When set to `remote`, the repository is cloned into " +
+				"a temporary directory and built. This is useful when the repository " +
+				"should act as the source of truth.",
```
[the remark this answers is not in the document]
Comment (mafredri) [reply options/options.go:444]: I felt it was more descriptive to have the two local/remote modes as opposed to simply a `--remote-repo-build-mode` flag. I can booleanize this though as I haven't planned for any additional modes.

[the remark this answers is not in the document]
Comment (johnstcn) [reply options/options.go:444]: Omakase!

File: envbuilder.go (hunk 4 above)
Comment (johnstcn) [reply envbuilder.go:146]: Does always cloning make the implementation simpler for now?

Comment (mafredri) [reply envbuilder.go:146]: Mostly it's just removal of the one if logic, but less things to go wrong so I suppose it is, yeah. Let's go for simplicity.

**Comment by mafredri** [comment]
Rebased on main and amended PR feedback, ready for round 2.

File: options/options.go [hunk 6]
```diff
@@ -482,6 +503,26 @@ func (o *Options) Markdown() string {
 	return sb.String()
 }
 
+func (o *Options) CABundle() ([]byte, error) {
```
Comment (johnstcn) [inline options/options.go:506]: I like this approach of hanging it off `Options`!

File: git/git_test.go [hunk 7]
```diff
@@ -169,6 +169,73 @@ func TestCloneRepo(t *testing.T) {
...
+				Username: "test",
+				Password: "test",
+			},
+		})
+		require.Error(t, err)
+	})
+	t.Run("OK", func(t *testing.T) {
+		// 2024/08/01 13:22:08 unsupported capability: shallow
+		// clone "http://127.0.0.1:41499": unexpected client error: unexpected requesting "http://127.0.0.1:41499/git-upload-pack" status code: 500
+		t.Skip("The gittest server doesn't support shallow cloning, skip for now...")
```
Comment (johnstcn) [inline git/git_test.go:208]: 😿

**Review by johnstcn — APPROVED** [review]
:shipit:

**Pull request merged**